garage/plotter/plotter.py

```diff
@@ -1,88 +1,153 @@
+                else:
+                    # Only fetch the last message of each type
+                    while not self._queue.empty():
+                        msg = self._queue.get_nowait()
+                        msgs[msg.op] = msg
+
+                if Op.STOP in msgs:
                     break
-            if 'stop' in msgs:
-                break
-            elif 'update' in msgs:
-                env, policy = msgs['update']
-                # env.start_viewer()
-            elif 'demo' in msgs:
-                param_values, max_length = msgs['demo']
-                policy.set_param_values(param_values)
-                rollout(
-                    env,
-                    policy,
-                    max_path_length=max_length,
-                    animated=True,
-                    speedup=5)
-            else:
-                if max_length:
+                elif Op.UPDATE in msgs:
+                    env, policy = msgs[Op.UPDATE].args
+                elif Op.DEMO in msgs:
+                    param_values, max_length = msgs[Op.DEMO].args
+                    policy.set_param_values(param_values)
+                    initial_rollout = False
                     rollout(
                         env,
                         policy,
                         max_path_length=max_length,
                         animated=True,
                         speedup=5)
-    except KeyboardInterrupt:
-        pass
-
-
-def _shutdown_worker():
-    if process:
-        queue.put(['stop'])
-        queue.close()
-        process.join()
-
-
-def init_worker():
-    global process, queue
-    queue = Queue()
-    process = Thread(target=_worker_start) if (
-        'Darwin' in platform.platform()) else Process(target=_worker_start)
-    process.daemon = True
-    process.start()
-    atexit.register(_shutdown_worker)
-
-
-def init_plot(env, policy):
-    global process, queue
-    if not (process and queue):
-        init_worker()
-
-    # Needed in order to draw glfw window on the main thread
-    if ('Darwin' in platform.platform()):
-        rollout(env, policy, max_path_length=np.inf, animated=True, speedup=5)
-
-    queue.put(['update', env, policy])
+                else:
+                    if max_length:
+                        rollout(
+                            env,
+                            policy,
+                            max_path_length=max_length,
+                            animated=True,
+                            speedup=5)
+        except KeyboardInterrupt:
+            pass
+
+    def shutdown(self):
+        if self._process.is_alive():
+            self._queue.put(Message(op=Op.STOP, args=None, kwargs=None))
+            self._queue.close()
+            self._process.join()
+
+    @staticmethod
+    def set_enable(enable):
+        """Set the plot enabling according to the run_experiment function.
+        """
+        Plotter.enable = enable
+
+    @staticmethod
+    def status(plot_algo_status):
```

Comment:
This function having an argument is a leaky abstraction. Either the Plotter is in charge of its enabled status, or the caller is. Not both.